### PR TITLE
[master] Initial commit of formula tests

### PR DESCRIPTION
**Pull request by techhat**
### What does this PR do?
These tests use actual formulas from github in an effort to make the test suite more well rounded. The formulas used are [docker](https://github.com/saltstack-formulas/docker-formula), [nginx](https://github.com/saltstack-formulas/nginx-formula), [users](https://github.com/saltstack-formulas/users-formula), and [salt](https://github.com/saltstack-formulas/salt-formula).

### What issues does this PR fix or reference?
Ref #64614, #64611, #64612, #64614.

### Merge requirements satisfied?
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by whytewolf — DISMISSED** [review]

File: tests/pytests/functional/formulas/test_docker.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+import pytest
+import requests
+
+
+@pytest.fixture(scope="module")
+def formula_tag():
+    return "2.4.2"
+
+
```
Comment (s0undt3ch) [inline tests/pytests/functional/formulas/test_docker.py:17]:
```suggestion
@pytest.fixture(scope="module")
def modules(loaders):
    return loaders.modules
    
    
@pytest.fixture(scope="module")
def formula_tag():
    return "2.4.2"


```

File: tests/pytests/functional/formulas/test_docker.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+@pytest.fixture(scope="module")
+def repo_url(formula_tag):
+    return f"https://github.com/saltstack-formulas/docker-formula/archive/refs/tags/v{formula_tag}.zip"
+
+
+@pytest.fixture(scope="module")
```
Comment (s0undt3ch) [inline tests/pytests/functional/formulas/test_docker.py:23]:
```suggestion
@pytest.fixture(scope="module", autouse=True)
```

File: tests/pytests/functional/formulas/test_docker.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        "test_|-docker-software-package-install-other_|-docker-software-package-install-other_|-show_notification",
+    ]
+    for state_id in state_ids:
+        assert ret["local"][state_id]["result"] is not False
+
+    state_ids = [
+        "test_|-docker-software-service-running-docker-fail-notify_|-docker-software-service-running-docker-fail-notify_|-fail_without_changes",
+    ]
+    for state_id in state_ids:
+        assert ret["local"][state_id]["result"] is False
```
Comment (s0undt3ch) [inline tests/pytests/functional/formulas/test_docker.py:82]:
```suggestion
def test_docker_formula(modules):
    ret = modules.state.sls("docker", test=True)
    for staterun in ret:
        if staterun.state_id == "test_|-docker-software-service-running-docker-fail-notify_|-docker-software-service-running-docker-fail-notify_|-fail_without_changes":
            assert staterun.result is False
        else:
            assert staterun.result is True
```
The returned object will be [MultiStateResult](https://github.com/saltstack/pytest-salt-factories/blob/1.0.0rc21/src/saltfactories/utils/functional.py#L463) for `state.sls` and each entry will be an instance of [`StateResult`](https://github.com/saltstack/pytest-salt-factories/blob/1.0.0rc21/src/saltfactories/utils/functional.py#L329)

Comment (s0undt3ch) [reply tests/pytests/functional/formulas/test_docker.py:82]: This way we're not going out to a shell, which we shouldn't, unless that's the purpose of the test.

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/functional/formulas/test_docker.py (hunk 3 above)
Comment (s0undt3ch) [reply tests/pytests/functional/formulas/test_docker.py:82]: Of course, this doesn't just apply to this test module, but all the ones you provided.

**Review by Ch3LL — DISMISSED** [review]

**Review by whytewolf — DISMISSED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Review by whytewolf — DISMISSED** [review]

File: tests/pytests/functional/formulas/test_docker.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        yield formula
+
+
+@pytest.fixture(scope="module")
+def modules(loaders, _formula):
+    return loaders.modules
+
+
+def test_docker_formula(modules):
+    ret = modules.state.sls("docker", test=True)
```
Comment (Ch3LL) [inline tests/pytests/functional/formulas/test_docker.py:19]: Why are we running all of these formulas with `test=True`?

We want the formula's to run to ensure the states run properly.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: tests/pytests/functional/formulas/test_docker.py (hunk 4 above)
Comment (garethgreenaway) [reply tests/pytests/functional/formulas/test_docker.py:19]: Looks like Salt is not finding the downloaded formulas:
`MultiStateResult(raw=["No matching sls found for 'docker' in env 'base'"], _structured=[])`

**Comment by Ch3LL** [comment]
Thanks for removing `test=True`. 

Just a follow up to @garethgreenaway 's comment here: https://github.com/saltstack/salt/pull/64677#discussion_r1296248831

Since the original tests did not actually catch that we weren't actually calling the correct formulas I think we need to improve our asserts in these tests because these tests are passing on linux hosts.

**Review by s0undt3ch — APPROVED** [review]

**Pull request merged**